text/0459-angle-bracket-built-in-components.md

```diff
@@ -0,0 +1,257 @@
+
+    <LinkTo @route="post.comment" @models={{array post comment}}>
+      Comment by {{comment.author.name}} on {{comment.date}}
+    </LinkTo>
+
+  {{/each}}
+{{/let}}
+```
+
+The singular `@model` argument is a special case of `@models`, provided as a
+convenience for the common case. Passing both `@model` and `@models` will be an
+error. (So would passing insufficient amount of models for the given route, as
+it already is today.)
+
+It is also possible to pass query params to the `{{link-to}}` component with
+the somewhat awkward `(query-params)` API. We propose to replace it with a
+`@query` named argument that simply take a regular hash (or POJO):
+
+```hbs
+{{#link-to "posts" (query-params direction="desc" showArchived=false)}}
+  Recent Posts
+{{/link-to}}
+
+...becomes...
+
+<LinkTo @route="posts" @query={{hash direction="desc" showArchived=false}}>
+  Recent Posts
+</LinkTo>
+```
+
+Finally, as mentioned above, `{{link-to}}` supports an "inline" form without a
+block. This form doesn't bring much value and used to cause confusion around
+the ordering of the arguments. We propose to simply not support this for the
+angle bracket invocation style:
+
+```hbs
+{{link-to "About Us" "about"}}
+
+...becomes...
+
+<LinkTo @route="about">About Us</LinkTo>
+```
+
+Other APIs of this compoment are already based on named arguments.
+
+#### Migration Path
+
+We would provide a codemod to convert the old invocation style into the new
+style.
+
+#### Deprecations
+
+Even though the angle bracket invocation style is recommended going forward
+(and should be enforced via a template lint), components can generally be
+invoked using the either the curly or angle bracket syntax. Therefore, while
+not recommended, `{{link-to}}` would still work and invoke the same component.
+
+In the case that the curly invocation style is used, we propose to issue the
+following deprecation warnings:
+
+* Passing positional arguments to `{{link-to}}` (pass named arguments instead)
+* Using the inline form of `{{link-to}}` (use the block form instead)
+* The `query-params` helper (pass a hash/POJO to the `query` argument instead)
+
+### `<Input>`
+
+Today, the `{{input}}` component is internally implemented as several internal
+components that are selected based on the `type` argument. This is intended as
+an internal implmentation detail, but as a result, it is not possible to invoke
+the component with `<Input>` since it does not exist as a "real" component.
+
+We propose to change this internal implementation strategy to make it possible
+to invoke this with angle brackets just like any other components.
+
+For example:
+
+```hbs
+{{input type="text" value=this.model.name}}
+
+...becomes...
+
+<Input @type="text" @value={{this.model.name}} />
+```
+
+Another example:
+
+```hbs
+{{input type="checkbox" name="email-opt-in" checked=this.model.emailPreference}}
+
+...becomes...
+
+<Input @type="checkbox" @name="email-opt-in" @checked={{this.model.emailPreference}} />
```

Comment:
If we were building these components from first principals, would we choose this API? Might this be an opportunity to build something better, removing `type` and the overloaded input API and instead provide dedicated components such as `<TextInput />` and `<Checkbox />`? (with possibly better names)

Reply:
I think there are advantages sticking close to underlying HTML element that it is trying to bridge (even though we don't fully support all the `input` types too). Also, this RFC just tries to fix a narrow problem (provide a 1:1 transition path from the curly invocations) without opening the can of worms of "hey.. what if we also just fix this _one_ thing......", they can be explored in separate RFCs IMO, so long as they are codemod-able I don't think there is a lot of extra cost to do them later.

Reply:
👍 thanks, agreed. Having the opportunity to codemod our way to some possible future API means that this is a nice incremental improvement

Reply:
> If we were building these components from first principals, would we choose this API? 

No, but it is not this APIs intent to "fix the design of the underlying components", that should be done separately.

tldr; what Godfrey said 😸

Reply:
This is the same point I came here to make. In particular, given `...attributes` and the semantics of `type` vs. `@type` on a regular `<input>`, it seems like it might be worth deprecating `{{input}}` (along the lines proposed here) and using the internal mechanics which create the correct input type to offer good suggestions. "You attempted to create a checkbox with `{{input}}`; we recommend `<CheckBox />`..." etc.

Reply:
I think both could work as @rwjblue points out this RFC is to bridge the gap with the commonly used components today. For instance if a Ember dev reads the guide materials on curly vs angle bracket and sees `{{input type="number"}}` being used it's a much smoother path to change it to `<Input @type="number">` vs having to know `<TextInput type="number" />` or all of the different variations.

A more angle bracket 1st set of input components would be great! We could even use codemods or build time AST transforms to rewrite to these new components in the future. But, for now let's make a way to use the built-in components we have.